From 4593723194675087ae98ea068b847fd3cc2ee9ce Mon Sep 17 00:00:00 2001
From: Etienne Cordonnier <ecordonnier@snap.com>
Date: Fri, 17 Mar 2023 10:33:11 +0100
Subject: [PATCH] android-tools: modifications to make it build in yocto

- Fix relocation errors on aarch64, e.g.:
"relocation R_AARCH64_ADR_PREL_PG_HI21 against symbol `_ZTV19SparseFileBufSource' which may bind externally can not be used when making a shared object; recompile with -fPIC"

- fix missing headers

Signed-off-by: Etienne Cordonnier <ecordonnier@snap.com>
---
 debian/system/append2simg.mk                    |  2 +-
 debian/system/fastboot.mk                       |  2 +-
 debian/system/img2simg.mk                       |  2 +-
 debian/system/libbacktrace.mk                   |  4 +++-
 debian/system/libbase.mk                        |  2 ++
 debian/system/libcutils.mk                      |  3 ++-
 debian/system/liblog.mk                         |  2 ++
 debian/system/libsparse.mk                      |  3 ++-
 debian/system/libutils.mk                       |  2 +-
 debian/system/libziparchive.mk                  |  3 ++-
 debian/system/simg2img.mk                       |  2 +-
 debian/system/simg2simg.mk                      |  2 +-
 frameworks/native/libs/adbd_auth/adbd_auth.cpp  |  2 ++
 packages/modules/adb/adb_listeners.cpp          | 12 ++++++------
 packages/modules/adb/client/transport_local.cpp |  2 +-
 15 files changed, 28 insertions(+), 17 deletions(-)

diff --git a/debian/system/append2simg.mk b/debian/system/append2simg.mk
index 5c5912fa..098cea99 100644
--- a/debian/system/append2simg.mk
+++ b/debian/system/append2simg.mk
@@ -11,7 +11,7 @@ CPPFLAGS += \
 
 LDFLAGS += \
   -Ldebian/out/system \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN/../lib/android' \
   -lbase \
   -llog \
   -lpthread \
diff --git a/debian/system/fastboot.mk b/debian/system/fastboot.mk
index b57bfdff..a05fc1a9 100644
--- a/debian/system/fastboot.mk
+++ b/debian/system/fastboot.mk
@@ -53,7 +53,7 @@ CPPFLAGS += \
 LDFLAGS += \
   -Ldebian/out/system \
   -L/usr/lib/$(DEB_HOST_MULTIARCH)/android \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN/../lib/android' \
   -lbase \
   -lcrypto \
   -lcutils \
diff --git a/debian/system/img2simg.mk b/debian/system/img2simg.mk
index df7d395d..87eff63b 100644
--- a/debian/system/img2simg.mk
+++ b/debian/system/img2simg.mk
@@ -11,7 +11,7 @@ CPPFLAGS += \
 
 LDFLAGS += \
   -Ldebian/out/system \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN/../lib/android' \
   -lbase \
   -llog \
   -lpthread \
diff --git a/debian/system/libbacktrace.mk b/debian/system/libbacktrace.mk
index 43ec216f..6fc1d876 100644
--- a/debian/system/libbacktrace.mk
+++ b/debian/system/libbacktrace.mk
@@ -1,3 +1,4 @@
+include rules_yocto.mk
 include /usr/share/dpkg/architecture.mk
 
 NAME = libbacktrace
@@ -77,7 +78,7 @@ CPPFLAGS += \
 LDFLAGS += \
   -Ldebian/out/external \
   -Ldebian/out/system \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN' \
   -Wl,-soname,$(NAME).so.0 \
   -lbase \
   -llog \
@@ -92,6 +93,7 @@ ifneq ($(filter armel mipsel,$(DEB_HOST_ARCH)),)
 endif
 
 build: $(OBJECTS_CXX) $(OBJECTS_ASSEMBLY)
+	mkdir -p debian/out/system/core
 	$(CXX) $^ -o debian/out/system/$(NAME).so.0 $(LDFLAGS)
 	ln -sf $(NAME).so.0 debian/out/system/$(NAME).so
 
diff --git a/debian/system/libbase.mk b/debian/system/libbase.mk
index b1e653bc..d1c1c16a 100644
--- a/debian/system/libbase.mk
+++ b/debian/system/libbase.mk
@@ -1,3 +1,4 @@
+include rules_yocto.mk
 NAME = libbase
 
 SOURCES = \
@@ -42,6 +43,7 @@ CPPFLAGS += \
 LDFLAGS += \
   -Ldebian/out/system \
   -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN' \
   -Wl,-soname,$(NAME).so.0 \
   -llog \
   -lpthread \
diff --git a/debian/system/libcutils.mk b/debian/system/libcutils.mk
index 6b789762..2e5b361f 100644
--- a/debian/system/libcutils.mk
+++ b/debian/system/libcutils.mk
@@ -1,3 +1,4 @@
+include rules_yocto.mk
 NAME = libcutils
 
 # system/core/libcutils/Android.bp
@@ -52,7 +53,7 @@ CPPFLAGS += \
 
 LDFLAGS += \
   -Ldebian/out/system \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN' \
   -Wl,-soname,$(NAME).so.0 \
   -lbase \
   -llog \
diff --git a/debian/system/liblog.mk b/debian/system/liblog.mk
index 3d2d50f9..d051a349 100644
--- a/debian/system/liblog.mk
+++ b/debian/system/liblog.mk
@@ -1,3 +1,4 @@
+include rules_yocto.mk
 NAME = liblog
 
 # system/logging/liblog/Android.bp
@@ -38,6 +39,7 @@ ifneq ($(filter armel mipsel,$(DEB_HOST_ARCH)),)
 endif
 
 build: $(OBJECTS)
+	mkdir -p debian/out/system
 	$(CXX) $^ -o debian/out/system/$(NAME).so.0 $(LDFLAGS)
 	ln -sf $(NAME).so.0 debian/out/system/$(NAME).so
 
diff --git a/debian/system/libsparse.mk b/debian/system/libsparse.mk
index 24d82183..eef1daaa 100644
--- a/debian/system/libsparse.mk
+++ b/debian/system/libsparse.mk
@@ -1,3 +1,4 @@
+include rules_yocto.mk
 NAME = libsparse
 
 # system/core/libsparse/Android.bp
@@ -20,7 +21,7 @@ CPPFLAGS += \
 
 LDFLAGS += \
   -Ldebian/out/system \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN' \
   -Wl,-soname,$(NAME).so.0 \
   -lbase \
   -lz \
diff --git a/debian/system/libutils.mk b/debian/system/libutils.mk
index ca8583a7..817b3974 100644
--- a/debian/system/libutils.mk
+++ b/debian/system/libutils.mk
@@ -43,7 +43,7 @@ CPPFLAGS += \
 
 LDFLAGS += \
   -Ldebian/out/system \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN' \
   -Wl,-soname,$(NAME).so.0 \
   -lbacktrace \
   -lcutils \
diff --git a/debian/system/libziparchive.mk b/debian/system/libziparchive.mk
index 0abfed8e..9d5ba2f7 100644
--- a/debian/system/libziparchive.mk
+++ b/debian/system/libziparchive.mk
@@ -1,3 +1,4 @@
+include rules_yocto.mk
 NAME = libziparchive
 
 # system/libziparchive/Android.bp
@@ -27,7 +28,7 @@ CPPFLAGS += \
 
 LDFLAGS += \
   -Ldebian/out/system \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN' \
   -Wl,-soname,$(NAME).so.0 \
   -lbase \
   -llog \
diff --git a/debian/system/simg2img.mk b/debian/system/simg2img.mk
index f8369e47..3ecee138 100644
--- a/debian/system/simg2img.mk
+++ b/debian/system/simg2img.mk
@@ -13,7 +13,7 @@ CPPFLAGS += \
 
 LDFLAGS += \
   -Ldebian/out/system \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN/../lib/android' \
   -lbase \
   -llog \
   -lpthread \
diff --git a/debian/system/simg2simg.mk b/debian/system/simg2simg.mk
index 69b74928..47d7a3ab 100644
--- a/debian/system/simg2simg.mk
+++ b/debian/system/simg2simg.mk
@@ -13,7 +13,7 @@ CPPFLAGS += \
 
 LDFLAGS += \
   -Ldebian/out/system \
-  -Wl,-rpath=/usr/lib/$(DEB_HOST_MULTIARCH)/android \
+  -Wl,-rpath='$$ORIGIN/../lib/android' \
   -lbase \
   -llog \
   -lpthread \
diff --git a/frameworks/native/libs/adbd_auth/adbd_auth.cpp b/frameworks/native/libs/adbd_auth/adbd_auth.cpp
index 15bd5c39..ebc74fb3 100644
--- a/frameworks/native/libs/adbd_auth/adbd_auth.cpp
+++ b/frameworks/native/libs/adbd_auth/adbd_auth.cpp
@@ -23,8 +23,10 @@
 #include <sys/eventfd.h>
 #include <sys/uio.h>
 
+#include <atomic>
 #include <chrono>
 #include <deque>
+#include <optional>
 #include <string>
 #include <string_view>
 #include <tuple>
diff --git a/packages/modules/adb/adb_listeners.cpp b/packages/modules/adb/adb_listeners.cpp
index 124e2d8d..88fb75dc 100644
--- a/packages/modules/adb/adb_listeners.cpp
+++ b/packages/modules/adb/adb_listeners.cpp
@@ -111,7 +111,7 @@ static void listener_event_func(int _fd, unsigned ev, void* _l)
 }
 
 // Called as a transport disconnect function. |arg| is the raw alistener*.
-static void listener_disconnect(void* arg, atransport*) EXCLUDES(listener_list_mutex) {
+static void listener_disconnect(void* arg, atransport*) {
     std::lock_guard<std::mutex> lock(listener_list_mutex);
     for (auto iter = listener_list.begin(); iter != listener_list.end(); ++iter) {
         if (iter->get() == arg) {
@@ -123,7 +123,7 @@ static void listener_disconnect(void* arg, atransport*) EXCLUDES(listener_list_m
 }
 
 // Write the list of current listeners (network redirections) into a string.
-std::string format_listeners() EXCLUDES(listener_list_mutex) {
+std::string format_listeners() {
     std::lock_guard<std::mutex> lock(listener_list_mutex);
     std::string result;
     for (auto& l : listener_list) {
@@ -142,7 +142,7 @@ std::string format_listeners() EXCLUDES(listener_list_mutex) {
 }
 
 InstallStatus remove_listener(const char* local_name, atransport* transport)
-    EXCLUDES(listener_list_mutex) {
+{
     std::lock_guard<std::mutex> lock(listener_list_mutex);
     for (auto iter = listener_list.begin(); iter != listener_list.end(); ++iter) {
         if (local_name == (*iter)->local_name) {
@@ -153,7 +153,7 @@ InstallStatus remove_listener(const char* local_name, atransport* transport)
     return INSTALL_STATUS_LISTENER_NOT_FOUND;
 }
 
-void remove_all_listeners() EXCLUDES(listener_list_mutex) {
+void remove_all_listeners() {
     std::lock_guard<std::mutex> lock(listener_list_mutex);
     auto iter = listener_list.begin();
     while (iter != listener_list.end()) {
@@ -176,7 +176,7 @@ void enable_server_sockets() EXCLUDES(listener_list_mutex) {
 }
 
 #if ADB_HOST
-void close_smartsockets() EXCLUDES(listener_list_mutex) {
+void close_smartsockets() {
     std::lock_guard<std::mutex> lock(listener_list_mutex);
     auto pred = [](const std::unique_ptr<alistener>& listener) {
         return listener->local_name == "*smartsocket*";
@@ -187,7 +187,7 @@ void close_smartsockets() EXCLUDES(listener_list_mutex) {
 
 InstallStatus install_listener(const std::string& local_name, const char* connect_to,
                                atransport* transport, int flags, int* resolved_tcp_port,
-                               std::string* error) EXCLUDES(listener_list_mutex) {
+                               std::string* error) {
     std::lock_guard<std::mutex> lock(listener_list_mutex);
     for (auto& l : listener_list) {
         if (local_name == l->local_name) {
diff --git a/packages/modules/adb/client/transport_local.cpp b/packages/modules/adb/client/transport_local.cpp
index 15a07246..c4e5f7f3 100644
--- a/packages/modules/adb/client/transport_local.cpp
+++ b/packages/modules/adb/client/transport_local.cpp
@@ -260,7 +260,7 @@ struct EmulatorConnection : public FdConnection {
 
 /* Only call this function if you already hold local_transports_lock. */
 static atransport* find_emulator_transport_by_adb_port_locked(int adb_port)
-        REQUIRES(local_transports_lock) {
+{
     auto it = local_transports.find(adb_port);
     if (it == local_transports.end()) {
         return nullptr;
-- 
2.36.1.vfs.0.0

